File: Épreuve_2_—_multi-lang/Java/Bracket.java
import java.util.Stack;

import static java.lang.System.exit;
import static java.lang.System.out;

public final class Bracket {

    private static final Stack<Character> characters = new Stack<>();

    public static void main(final String[] args) {
        final char[] chars = args[0].toCharArray();
        for (final char c : chars)
            switch (c) {
                case '(':
                    characters.push('(');
                    break;
                case ')':
                    checkBracket('(');
                    break;
                case '{':
                    characters.push('{');
                    break;
                case '}':
                    checkBracket('{');
                    break;
                case '[':
                    characters.push('[');
                    break;
                case ']':
                    checkBracket('[');
                    break;
            }
        if (characters.isEmpty())
            out.println("YES");
        else
            out.println("NO");
    }

    private static void checkBracket(final char bracket) {
        if (characters.pop() != bracket) {
            out.println("NO");
            exit(0);
        }
    }
}

File: Épreuve_2_—_multi-lang/Kotlin/Bracket.kt
import java.lang.System.out
import java.util.*

object Bracket {

    private val OPEN_BRACKETS = Arrays.asList('(', '{', '[')
    private val openBrackets = Stack<Char>()
    private val CLOSE_BRACKETS = Arrays.asList(')', '}', ']')
    private val closeBrackets = ArrayDeque<Char>()

    @JvmStatic
    fun main(args: Array<String>) {
        out.println(if (checkBracket(args[0].toCharArray())) "YES" else "NO")
    }

    private fun checkBracket(chars: CharArray): Boolean {
        for (c in chars)
            if (OPEN_BRACKETS.contains(c))
                openBrackets.push(c)
            else if (CLOSE_BRACKETS.contains(c))
                closeBrackets.offer(c)

        if (openBrackets.size != closeBrackets.size)
            return false

        while (!openBrackets.isEmpty())
            if (closeBrackets.poll() != CLOSE_BRACKETS[OPEN_BRACKETS.indexOf(openBrackets.pop())])
                return false

        return true
    }
}
